apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1/HiPerfSubscriber.cs

```diff
@@ -0,0 +1,723 @@
+                    }
+                    catch (Exception e) when (e.IsCancellation() || e.IsRpcCancellation() || (e.As<RpcException>()?.IsRecoverable() ?? false))
+                    {
+                        // If it's not an unrecoverable error, then allow non-exceptional code-paths to handle. Do nothing here.
+                    }
+                    finally
+                    {
+                        // Try to cleanly shutdown the subscriber. This may or may not be possible
+                        // depending on the state of the subscriber; ignore all errors.
+                        if (pull != null)
+                        {
+                            _registerTaskFn(_taskHelper.Run(() => pull.WriteCompleteAsync()));
+                        }
+                    }
+                }
+            }
+
+            private async Task Pull(SubscriberClient.StreamingPullStream pull, CancellationToken streamingPullToken)
+            {
+                var cts = CancellationTokenSource.CreateLinkedTokenSource(streamingPullToken, _softStopCts.Token);
+                while (!cts.IsCancellationRequested && await _taskHelper.ConfigureAwait(pull.ResponseStream.MoveNext(CancellationToken.None)))
+                {
+                    // If in soft-stop, the overall StreamingPull might not yet be cancelled, but don't process any more msgs.
+                    if (cts.IsCancellationRequested)
+                    {
+                        break;
+                    }
+                    var msgs = pull.ResponseStream.Current.ReceivedMessages;
+                    var ackIds = new HashSet<string>(msgs.Select(x => x.AckId));
+                    var allMsgsHandledCts = CancellationTokenSource.CreateLinkedTokenSource(_hardStopCts.Token);
+                    _registerTaskFn(_taskHelper.Run(() => AutoExtend(ackIds, allMsgsHandledCts.Token)));
+                    foreach (var msg in msgs)
+                    {
+                        if (cts.IsCancellationRequested)
+                        {
+                            break;
+                        }
+                        // Returned Task completes once flow-control is satisfied.
+                        await _taskHelper.ConfigureAwait(_flow.Process(msg.CalculateSize(), async () =>
+                        {
+                            if (cts.IsCancellationRequested)
+                            {
+                                return;
+                            }
+                            lock (_qLock)
+                            {
+                                _unAckedMsgCount += 1;
+                            }
+                            var reply = await _taskHelper.ConfigureAwaitHideErrors(_handlerAsync(msg.Message, _hardStopCts.Token), Reply.Nack);
+                            TaskCompletionSource<int> qEvent = null;
+                            lock (_qLock)
+                            {
+                                if (reply == Reply.Ack)
```

Comment:
Yes. I don't think any other behaviour makes sense.
The only extra thing to do would be to log it, but logging is still TODO.